ships/0028-buildrun-cleanup.md

```diff
@@ -0,0 +1,152 @@
+<!--
+Copyright The Shipwright Contributors
+
+SPDX-License-Identifier: Apache-2.0
+-->
+
+---
+title: buildrun-cleanup
+authors:
+  - @SaschaSchwarze0
+reviewers:
+  - @HeavyWombat
+  - @GabeMontero
+approvers:
+  - @adambkaplan
+  - @qu1queee
+creation-date: 2022-02-14
+last-updated: 2022-02-14
+status: partially implementable
+---
+
+# BuildStrategy Parameter Enhancements
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [x] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [docs](/docs/)
+
+## Open Questions [optional]
+
+See TBDs
+
+## Summary
+
+This ship proposes means for Build users to define that their completed BuildRuns should be automatically cleaned up after some time or when a certain number of BuildRuns is reached.
+
+## Motivation
+
+In Shipwright, we separated the definition of a build and the run of a build into two artifacts, the Build and BuildRun. The BuildRun by its nature is a run-to-completion workload. Once it is completed, its only purpose is to view its status. Retaining the BuildRun for some time makes sense. In case of a successful BuildRun on finds there the digest of the image. For a failed BuildRun, one finds the error details and a pointer to the pod logs. The lifecycle of the BuildRun is coupled with the TaskRun and Pod using [owner references](https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/).
+
+Over times, the number of BuildRuns in the system grows. This consumes resources in Kubernetes' database (usually etcd) and on the nodes (the pod logs).
+
+Eventually, users must manually clean up old BuildRuns which is not nice. Providing them means to define that BuildRuns get automatically cleaned up simplifies their lives.
+
+### Goals
+
+- Allow Build user to define how many completed BuildRuns should be kept.
+- Allow Build user to define how long a completed BuildRun should be kept.
+- Allow Build user to separate successful and failed BuildRuns in their cleanup handling.
+
+### Non-Goals
+
+- Allowing BuildRuns to be archived to an external database.
+
+## Proposal
+
+### API Changes
+
+We will extend the Build resource with an additional optional `retention` section`:
+
+```yaml
+spec:
+  retention:
+    ttlAfterFailed: duration
+    ttlAfterSuccessful: duration
```

Comment:
on the wording, `successful` vs `succeeded` . Succeeded inlines with how we do in the `.status.conditions`, see https://github.com/shipwright-io/build/blob/main/docs/buildrun.md#understanding-the-state-of-a-buildrun

Reply:
Done.